sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -477,7 +477,10 @@ case class Add(
   override protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression): Add =
     copy(left = newLeft, right = newRight)
 
-  override lazy val canonicalized: Expression = {
+  override lazy val canonicalized: Expression = dataType match {
+    case _: DecimalType =>
```

Comment:
can we add some comments to explain the reason?

Reply:
thank you @gengliangwang for the catching. 
can we make it more fine-grained ? Not all decimal add will fail, so we can check if we can reorder them safely. e.g., precision and scale in all left and right are same.

Reply:
@cloud-fan comment added

Reply:
@ulysses-you  Are you sure about that? My concern is that if both left and right contains integer contains decimal `Add`s, the result may still be different after sorting all the sub `Add`s

Reply:
How about adding an extra `Cast` into the canonicalized form if needed like:
```
  override lazy val canonicalized: Expression = {
    // TODO: do not reorder consecutive `Add`s with different `evalMode`
    val reordered = orderCommutative({ case Add(l, r, _) => Seq(l, r) }).reduce(Add(_, _, evalMode))
    if (dataType != reordered.dataType) {
      Cast(reordered, dataType)
    } else {
      reordered
    }
  }
```

Reply:
Hmm, maybe adding an extra `Cast` is not a good idea as the 2 expressions with different `dataType`s shouldn't be considered equal.

Reply:
@peter-toth The ideal solution would be adding extra casts in all the canonicalization of the children of`ComplexTypeMergingExpression` if the data type is changed. However, there are also overriding in some of the `ComplexTypeMergingExpression`.
So I would take your suggestion to reorder the Add if the result data type is not changed. Thank you.